lib/Optimizer/GraphOptimizer/PassManager.cpp

```diff
@@ -0,0 +1,163 @@
+
+#include "llvm/Support/CommandLine.h"
+
+#include <glog/logging.h>
+
+using namespace glow;
+
+namespace {
+llvm::cl::OptionCategory passManagerCat("PassManager Options");
+
+llvm::cl::opt<bool> verifyBeforeAllPassesOpt(
+    "verify-before-all-passes",
+    llvm::cl::desc("Verify the Function before all passes."),
+    llvm::cl::Optional, llvm::cl::cat(passManagerCat));
+
+llvm::cl::list<std::string> verifyBeforePassesOpt(
+    "verify-before-passes",
+    llvm::cl::desc("Verify the Function before the listed Passes."),
+    llvm::cl::value_desc("FunctionPass names (e.g. DCE,CSE)"),
+    llvm::cl::ZeroOrMore, llvm::cl::CommaSeparated,
+    llvm::cl::cat(passManagerCat));
+
+llvm::cl::opt<bool> verifyAfterAllPassesOpt(
+    "verify-after-all-passes",
+    llvm::cl::desc("Verify the Function after all passes."), llvm::cl::Optional,
+    llvm::cl::cat(passManagerCat));
+
+llvm::cl::list<std::string> verifyAfterPassesOpt(
+    "verify-after-passes",
+    llvm::cl::desc("Verify the Function after the listed Passes."),
+    llvm::cl::value_desc("FunctionPass names (e.g. DCE,CSE)"),
+    llvm::cl::ZeroOrMore, llvm::cl::CommaSeparated,
+    llvm::cl::cat(passManagerCat));
+
+llvm::cl::opt<bool> dumpDAGBeforeAllPassesOpt(
+    "dump-graph-DAG-before-all-passes",
+    llvm::cl::desc(
+        "Debug option to export the Graph in DOT format before all passes."),
+    llvm::cl::Optional, llvm::cl::cat(passManagerCat));
+
+llvm::cl::list<std::string> dumpDAGBeforePassesOpt(
+    "dump-graph-DAG-before-passes",
+    llvm::cl::desc("Debug option to export the Graph in DOT format before the "
+                   "listed Passes."),
+    llvm::cl::value_desc("FunctionPass names (e.g. DCE,CSE)"),
+    llvm::cl::ZeroOrMore, llvm::cl::CommaSeparated,
+    llvm::cl::cat(passManagerCat));
+
+llvm::cl::opt<bool> dumpDAGAfterAllPassesOpt(
+    "dump-graph-DAG-after-all-passes",
+    llvm::cl::desc(
+        "Debug option to export the Graph in DOT format after all passes."),
+    llvm::cl::Optional, llvm::cl::cat(passManagerCat));
+
+llvm::cl::list<std::string> dumpDAGAfterPassesOpt(
+    "dump-graph-DAG-after-passes",
+    llvm::cl::desc("Debug option to export the Graph in DOT format after the "
+                   "listed Passes."),
+    llvm::cl::value_desc("FunctionPass names (e.g. DCE,CSE)"),
+    llvm::cl::ZeroOrMore, llvm::cl::CommaSeparated,
+    llvm::cl::cat(passManagerCat));
+
+/// Helper to check if \p otherStr is in \p strList.
+static bool listContainsString(llvm::ArrayRef<std::string> strList,
+                               llvm::StringRef otherStr) {
+  for (llvm::StringRef str : strList) {
+    if (str == otherStr) {
+      return true;
+    }
+  }
+  return false;
+}
+} // namespace
+
+bool FunctionPassManager::runPrePass(const FunctionPass &P) {
+  if (dumpDAGBeforeAllPassesOpt ||
+      listContainsString(dumpDAGBeforePassesOpt, P.getName())) {
+    F_->dumpDAG(F_->getName().str() + "_PrePass_" + P.getName().str() + ".dot");
```

Comment:
If we run the same pass on the same function multiple times, this line would overwrite previous dumps and save only the last one. If would would maintain a pass number (inside PM or global) this could be easily solved by adding this pass number a suffix.

Reply:
Done.

Reply:
Cool! Thanks!

But now that we have a pass counter let's go a step further ;-) ;-)
- Let's add an option to print all the passes being run, e.g. `Pass #N: CSE`.  May be also the pass manager name. This gives us a the log of which passes were run an in which order.
- Let's add an option to stop at the pass number N. This would allow for bisection of the pass pipeline in case when one of the passes introduces an error.

Reply:
Done 🙂 

- `-print-passes`
- `-stop-passes-after-num=N`